From 199f07e5c0c744bd28c7038b403f9972c7ebfa9b Mon Sep 17 00:00:00 2001
From: Nik Everett <neverett@wikimedia.org>
Date: Thu, 16 Oct 2014 08:11:10 -0700
Subject: [PATCH] Only initialze sweepables on new threads

Closes #3999
---

diff --git a/hphp/runtime/base/sweepable.cpp b/hphp/runtime/base/sweepable.cpp
index a886abf..ca60b57 100644
--- a/hphp/runtime/base/sweepable.cpp
+++ b/hphp/runtime/base/sweepable.cpp
@@ -39,6 +39,7 @@ void Sweepable::Node::delist() {
   p->next = n;
 }
 
+// Called once per thread initialization from ThreadInfo::init
 void Sweepable::InitSweepableList() {
   t_sweep.init();
 }
diff --git a/hphp/runtime/base/thread-info.cpp b/hphp/runtime/base/thread-info.cpp
index 215bb96..0889648 100644
--- a/hphp/runtime/base/thread-info.cpp
+++ b/hphp/runtime/base/thread-info.cpp
@@ -65,6 +65,7 @@ ThreadInfo::ThreadInfo()
 
   Lock lock(s_thread_info_mutex);
   s_thread_infos.insert(this);
+  Sweepable::InitSweepableList();
 }
 
 ThreadInfo::~ThreadInfo() {
diff --git a/hphp/runtime/base/thread-init-fini.cpp b/hphp/runtime/base/thread-init-fini.cpp
index d685ab5..f533bf5 100644
--- a/hphp/runtime/base/thread-init-fini.cpp
+++ b/hphp/runtime/base/thread-init-fini.cpp
@@ -50,7 +50,6 @@ InitFiniNode::InitFiniNode(void(*f)(), When init) {
 }
 
 void init_thread_locals(void *arg /* = NULL */) {
-  Sweepable::InitSweepableList();
   ServerStats::GetLogger();
   zend_get_bigint_data();
   zend_get_rand_data();
